File: frontend/js/views/tax_bracket_optimizer.js
// Federal marginal tax-bracket optimizer.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';

export async function renderTaxBracketOptimizer(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.tax_bracket_optimizer.title">// TAX BRACKET OPTIMIZER</span></h1>
        <p class="muted small" data-i18n-html="view.tax_bracket_optimizer.intro">
            See how much room is left in your current marginal bracket and how much you
            can realise (Roth conversion, IRA withdrawal, LT cap gains) before bumping
            into the next rate. Embedded 2026 IRS brackets for single / MFJ / HoH.
            Effective rate &lt; marginal because of progressive tiering.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.tax_bracket_optimizer.field.status">Filing status</span>
                    <select id="tbo-status">
                        <option value="single" selected>Single</option>
                        <option value="mfj">Married filing jointly</option>
                        <option value="hoh">Head of household</option>
                    </select>
                </label>
                <label><span class="muted small" data-i18n="view.tax_bracket_optimizer.field.income">Taxable ordinary income $</span>
                    <input type="number" id="tbo-income" step="1000" min="0" value="80000" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="tbo-run" data-shortcut="r" data-i18n="view.tax_bracket_optimizer.btn.run">⚡ Compute</button>
            <div id="tbo-result"></div>
        </div>
    `;
    mount.querySelector('#tbo-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#tbo-result');
    const input = {
        filing_status: mount.querySelector('#tbo-status').value,
        taxable_ordinary_income_usd: parseFloat(mount.querySelector('#tbo-income').value) || 0,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.tax_bracket_optimizer.status.computing'))}</p>`;
    try {
        const r = await api.request('/tax-bracket-optimizer/compute', { method: 'POST', body: JSON.stringify(input) });
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.tax_bracket_optimizer.field.marginal'))}</div>
                    <strong style="font-size:1.4em">${r.current_marginal_rate_pct.toFixed(0)}%</strong></div>
                <div><div class="muted small">${esc(t('view.tax_bracket_optimizer.field.room'))}</div>
                    <strong class="pos">$${r.room_in_current_bracket_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.tax_bracket_optimizer.field.next_rate'))}</div>
                    <strong>${r.next_bracket_rate_pct.toFixed(0)}%</strong></div>
                <div><div class="muted small">${esc(t('view.tax_bracket_optimizer.field.tax'))}</div>
                    <strong class="neg">$${r.federal_tax_liability_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.tax_bracket_optimizer.field.effective'))}</div>
                    <strong>${r.effective_rate_pct.toFixed(2)}%</strong></div>
                <div><div class="muted small">${esc(t('view.tax_bracket_optimizer.field.bracket_upper'))}</div>
                    <strong>${isFinite(r.current_bracket_upper_usd) ? '$' + r.current_bracket_upper_usd.toFixed(0) : '∞'}</strong></div>
            </div>
            <h2 style="margin-top:1rem">${esc(t('view.tax_bracket_optimizer.h2.brackets'))}</h2>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.tax_bracket_optimizer.th.rate">Rate</th>
                    <th data-i18n="view.tax_bracket_optimizer.th.lower">Lower</th>
                    <th data-i18n="view.tax_bracket_optimizer.th.upper">Upper</th>
                </tr></thead>
                <tbody>${(r.brackets || []).map(b => {
                    const isCurrent = b.rate_pct === r.current_marginal_rate_pct;
                    return `<tr style="${isCurrent ? 'background:rgba(255,42,109,0.08)' : ''}">
                        <td>${b.rate_pct.toFixed(0)}%</td>
                        <td>$${b.lower_usd.toFixed(0)}</td>
                        <td>${isFinite(b.upper_usd) ? '$' + b.upper_usd.toFixed(0) : '∞'}</td>
                    </tr>`;
                }).join('')}</tbody>
            </table>
        `;
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

File: crates/traderview-db/src/tax_bracket_optimizer.rs
//! Federal marginal tax-bracket optimizer.
//!
//! Lets the user see how much "room" they have left in their current
//! marginal bracket, and how much they could realise (Roth conversion,
//! IRA withdrawal, capital gain harvest) before bumping into the next
//! bracket.
//!
//! Brackets are 2026 published IRS values for ordinary income, single
//! / married-filing-jointly / head-of-household. Standard deduction
//! 2026: single $15,000, MFJ $30,000, HoH $22,500 (per IRS inflation
//! adjustments).
//!
//! Pure compute — no tax-table lookups beyond the embedded values.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct TaxBracketInput {
    pub filing_status: String,  // "single" | "mfj" | "hoh"
    pub taxable_ordinary_income_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct BracketRow {
    pub rate_pct: f64,
    pub lower_usd: f64,
    pub upper_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct TaxBracketReport {
    pub filing_status: String,
    pub taxable_income_usd: f64,
    pub current_marginal_rate_pct: f64,
    pub current_bracket_upper_usd: f64,
    pub room_in_current_bracket_usd: f64,
    pub next_bracket_rate_pct: f64,
    pub federal_tax_liability_usd: f64,
    pub effective_rate_pct: f64,
    pub brackets: Vec<BracketRow>,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

/// Returns 2026 brackets for the given filing status as (rate%, lower).
/// Upper is implied by the next bracket's lower (last upper = infinity).
fn brackets_2026(status: &str) -> Vec<(f64, f64)> {
    match status {
        "mfj" => vec![
            (10.0, 0.0),
            (12.0, 23_850.0),
            (22.0, 96_950.0),
            (24.0, 206_700.0),
            (32.0, 394_600.0),
            (35.0, 501_050.0),
            (37.0, 751_600.0),
        ],
        "hoh" => vec![
            (10.0, 0.0),
            (12.0, 17_000.0),
            (22.0, 64_850.0),
            (24.0, 103_350.0),
            (32.0, 197_300.0),
            (35.0, 250_500.0),
            (37.0, 626_350.0),
        ],
        _ => vec![ // single (default)
            (10.0, 0.0),
            (12.0, 11_925.0),
            (22.0, 48_475.0),
            (24.0, 103_350.0),
            (32.0, 197_300.0),
            (35.0, 250_525.0),
            (37.0, 626_350.0),
        ],
    }
}

pub fn liability(brackets: &[(f64, f64)], taxable: f64) -> f64 {
    if taxable <= 0.0 { return 0.0; }
    let mut tax = 0.0_f64;
    for i in 0..brackets.len() {
        let (rate, lower) = brackets[i];
        let upper = brackets.get(i + 1).map(|b| b.1).unwrap_or(f64::INFINITY);
        if taxable <= lower { break; }
        let bracket_top = taxable.min(upper);
        tax += (bracket_top - lower) * rate / 100.0;
    }
    tax
}

pub fn compute(input: &TaxBracketInput) -> TaxBracketReport {
    let raw = brackets_2026(&input.filing_status);
    let taxable = input.taxable_ordinary_income_usd.max(0.0);

    // Find current bracket.
    let mut current_rate = raw[0].0;
    let mut current_upper = f64::INFINITY;
    let mut next_rate = raw[0].0;
    for i in 0..raw.len() {
        let (rate, lower) = raw[i];
        let upper = raw.get(i + 1).map(|b| b.1).unwrap_or(f64::INFINITY);
        if taxable >= lower && taxable < upper {
            current_rate = rate;
            current_upper = upper;
            next_rate = raw.get(i + 1).map(|b| b.0).unwrap_or(rate);
            break;
        }
    }
    let room = (current_upper - taxable).max(0.0);
    let liab = liability(&raw, taxable);
    let eff_rate = if taxable > 0.0 { liab / taxable * 100.0 } else { 0.0 };

    let brackets: Vec<BracketRow> = raw
        .iter()
        .enumerate()
        .map(|(i, (rate, lower))| BracketRow {
            rate_pct: *rate,
            lower_usd: *lower,
            upper_usd: raw.get(i + 1).map(|b| b.1).unwrap_or(f64::INFINITY),
        })
        .collect();

    TaxBracketReport {
        filing_status: input.filing_status.clone(),
        taxable_income_usd: taxable,
        current_marginal_rate_pct: current_rate,
        current_bracket_upper_usd: current_upper,
        room_in_current_bracket_usd: room,
        next_bracket_rate_pct: next_rate,
        federal_tax_liability_usd: liab,
        effective_rate_pct: eff_rate,
        brackets,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn brackets_2026_single_first_bracket() {
        let b = brackets_2026("single");
        assert_eq!(b[0], (10.0, 0.0));
        assert_eq!(b[1].0, 12.0);
        assert_eq!(b[1].1, 11_925.0);
    }

    #[test]
    fn brackets_2026_mfj_first_bracket() {
        let b = brackets_2026("mfj");
        assert_eq!(b[1].1, 23_850.0);
    }

    #[test]
    fn brackets_unknown_status_defaults_single() {
        let single = brackets_2026("single");
        let bogus = brackets_2026("bogus");
        assert_eq!(single, bogus);
    }

    #[test]
    fn liability_zero_income() {
        let b = brackets_2026("single");
        assert_eq!(liability(&b, 0.0), 0.0);
    }

    #[test]
    fn liability_only_first_bracket_single() {
        let b = brackets_2026("single");
        // $10k income, all at 10% = $1000
        let l = liability(&b, 10_000.0);
        assert!((l - 1_000.0).abs() < 0.5);
    }

    #[test]
    fn liability_spans_first_two_brackets() {
        let b = brackets_2026("single");
        // $20k taxable: first $11,925 @ 10% = $1192.50, next $8075 @ 12% = $969
        let l = liability(&b, 20_000.0);
        let expected = 11_925.0 * 0.10 + (20_000.0 - 11_925.0) * 0.12;
        assert!((l - expected).abs() < 0.5);
    }

    #[test]
    fn compute_current_marginal_single_first_bracket() {
        let r = compute(&TaxBracketInput {
            filing_status: "single".into(),
            taxable_ordinary_income_usd: 5_000.0,
        });
        assert_eq!(r.current_marginal_rate_pct, 10.0);
        assert_eq!(r.next_bracket_rate_pct, 12.0);
        // room = $11,925 − $5,000 = $6,925
        assert!((r.room_in_current_bracket_usd - 6_925.0).abs() < 0.5);
    }

    #[test]
    fn compute_current_marginal_22_pct_mid_bracket() {
        let r = compute(&TaxBracketInput {
            filing_status: "single".into(),
            taxable_ordinary_income_usd: 80_000.0,
        });
        // $80k single: bracket = 22% [48,475 → 103,350]
        assert_eq!(r.current_marginal_rate_pct, 22.0);
        // room = $103,350 − $80,000 = $23,350
        assert!((r.room_in_current_bracket_usd - 23_350.0).abs() < 0.5);
    }

    #[test]
    fn compute_mfj_lower_marginal_at_same_income() {
        let single = compute(&TaxBracketInput {
            filing_status: "single".into(), taxable_ordinary_income_usd: 100_000.0,
        });
        let mfj = compute(&TaxBracketInput {
            filing_status: "mfj".into(), taxable_ordinary_income_usd: 100_000.0,
        });
        // MFJ brackets wider, so $100k MFJ is in lower marginal than single.
        assert!(mfj.current_marginal_rate_pct <= single.current_marginal_rate_pct);
    }

    #[test]
    fn compute_effective_rate_below_marginal() {
        let r = compute(&TaxBracketInput {
            filing_status: "single".into(), taxable_ordinary_income_usd: 200_000.0,
        });
        // 32% marginal but effective is lower (progressive).
        assert!(r.effective_rate_pct < r.current_marginal_rate_pct);
    }

    #[test]
    fn compute_bracket_row_count_seven() {
        let r = compute(&TaxBracketInput {
            filing_status: "single".into(), taxable_ordinary_income_usd: 50_000.0,
        });
        assert_eq!(r.brackets.len(), 7);
    }

    #[test]
    fn compute_zero_income_zero_liability() {
        let r = compute(&TaxBracketInput {
            filing_status: "single".into(), taxable_ordinary_income_usd: 0.0,
        });
        assert_eq!(r.federal_tax_liability_usd, 0.0);
        assert_eq!(r.effective_rate_pct, 0.0);
    }
}
